plugins/gradle/resources/org/jetbrains/plugins/gradle/GradleImprovedHotswapDetectionInitScript.groovy

```diff
@@ -0,0 +1,137 @@
+ * e.g.
+ *   root: "/repo/project/build/classes/java/main/",
+ *   class: "com/acme/MyClass.class"
+ *
+ * Then, in
+ * {@link org.jetbrains.plugins.gradle.execution.build.GradleProjectTaskRunner#getAffectedOutputRoots}
+ * we read initScriptOutputFile and call
+ * {@link com.intellij.task.ProjectTaskContext#fileGenerated}
+ */
+
+Map<Task,Map<File,String>> beforeTaskFileHashes = [:]
+def effectiveTasks = []
+
+def initScriptOutputFile = new File("%s")
+
+gradle.taskGraph.addTaskExecutionListener(new TaskExecutionAdapter() {
+  void beforeExecute(Task task) {
+    if (task.outputs.hasOutput) {
+      beforeTaskFileHashes.put(task, hashOutputs(task.outputs.files.files))
+    }
+  }
+
+  void afterExecute(Task task, TaskState state) {
+    if ((state.didWork || (state.skipped && state.skipMessage == 'FROM-CACHE')) && task.outputs.hasOutput) {
+      effectiveTasks.add(task)
+    }
+  }
+})
+
+gradle.addBuildListener(new BuildAdapter() {
+  void buildFinished(BuildResult result) {
+    effectiveTasks.each { Task task ->
+      def beforeHashes = beforeTaskFileHashes[task]
+      task.outputs.files.files.each { File outputDir ->
+        if (outputDir.directory) {
+          def hashes = hashOutputDir(outputDir)
+          hashes.each { file, hash ->
+            def beforeHash = beforeHashes[file]
+            if (beforeHash == null || beforeHash != hash) {
+              def relativePath = outputDir.toPath().relativize(file.toPath())
+              logger.info "> Task: ${task.path} generated class '$relativePath' in '${outputDir.path}'"
+
+              // 'initScriptOutputFile' variable is declared in GradleProjectTaskRunner.java
+              initScriptOutputFile.append("root:${outputDir.path}\n")
+              initScriptOutputFile.append("path:${relativePath}\n")
+            }
+          }
+        }else {
+          def beforeHash = beforeHashes[outputDir]
+          def hash = hashFile(outputDir)
+          if (beforeHash == null || beforeHash != hash) {
+            initScriptOutputFile.append("root:${outputDir.path}\n")
+            initScriptOutputFile.append("path:\n")
+          }
+        }
+      }
+    }
+  }
+})
+
+private static Map<File,String> hashOutputs(Set<File> outputs) {
+  Map<File,String> hashes = [:]
+
+  outputs.each { outputDir ->
+    if (outputDir.isDirectory()) {
+      hashFilesInDir(outputDir, hashes)
+    }
+  }
+  return hashes
+}
+
+private static Map<File,String> hashOutputDir(File outputDir) {
+  Map<File,String> hashes = [:]
+  hashFilesInDir(outputDir, hashes)
+  return hashes
+}
+
+private static hashFilesInDir(File dir, Map<File,String> hashes) {
+  dir.eachFileRecurse(FILES) {
+    hashes.put(it, hashFile(it))
+  }
+}
+
+private static String hashFile(File file) {
+  // Set your algorithm
+  // "MD2","MD5","SHA","SHA-1","SHA-256","SHA-384","SHA-512"
+  MessageDigest hash = MessageDigest.getInstance("SHA-256")
+
+  FileInputStream fis = new FileInputStream(file)
+  try {
```

Comment:
this seems low level, does it need a try with resources for fis? there's something like that in groovy yeah? like `withReader` or something?

Reply:
Refactored using Groovy methods 👍